grove/simon/simon.py

```diff
@@ -0,0 +1,424 @@
+"""Module for the Simon's Algorithm.
+For more information, see
+
+- http://lapastillaroja.net/wp-content/uploads/2016/09/\
+Intro_to_QC_Vol_1_Loceff.pdf
+- http://pages.cs.wisc.edu/~dieter/Courses/2010f-CS880/Scribes/05/lecture05.pdf
+"""
+
+import numpy as np
+import pyquil.quil as pq
+from pyquil.gates import *
+
+
+def unitary_function(mappings):
+    """
+    Creates a unitary transformation that maps each state
+    to the values specified in mappings.
+
+    Some (but not all) of these transformations involve a scratch qubit,
+    so one is always provided. That is, if given the mapping of :math:`n`
+    qubits, the calculated transformation will be on :math:`n + 1` qubits,
+    where the zeroth qubit is the scratch bit and the return value
+    of the function is left in the qubits that follow.
+
+    :param list(int) mappings: List of the mappings of :math:`f(x)` on
+                               all length :math:`n` in their decimal
+                               representations.
+                               For example, the following mapping:
+
+                               - :math:`00 \\rightarrow 00`
+                               - :math:`01 \\rightarrow 10`
+                               - :math:`10 \\rightarrow 10`
+                               - :math:`11 \\rightarrow 00`
+
+                               Would be represented as :math:`[0, 2, 2, 0]`.
+                               Requires mappings to be either one-to-one,
+                               or two-to-one with unique mask :math:`s`,
+                               as specified in Simon's problem.
+    :return: Matrix representing specified unitary transformation.
+    :rtype: numpy array
+    """
+    assert len(mappings) > 1, \
+        "function domain must be at least one bit (size 2)"
+
+    n = int(np.log2(len(mappings)))
+    assert len(mappings) == 2 ** n, \
+        "mappings must have a length that is a power of two"
+
+    distinct_outputs = len(set(mappings))
+    assert distinct_outputs in {2 ** (n - 1), 2 ** n}, \
+        "Function must be one-to-one or two-to-one"
```

Comment:
Yup, you're right. I've made this check now part of the loop, and also added a check to make sure the mask is consistent throughout for two-to-one mappings. I've also changed this to raise ValueErrors rather than assert. 

To a similar point as your other comment, it may be worth moving these checks outside this method if unitary_function was made more general and in another module.